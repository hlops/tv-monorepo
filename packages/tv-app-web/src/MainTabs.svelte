<script>
  import { WiredTab } from "./ui/wired/wired-tab";
  import { WiredTabs } from "./ui/wired/wired-tabs";
  import TvGuidesUpload from "./ui/tvGuides/TvGuidesUpload.svelte";
  import Channels from "./ui/channles/Channels.svelte";
  import TvGuides from "./ui/tvGuides/TvGuides.svelte";
</script>

<wired-tabs selected="Two">
  <wired-tab name="One" label="Каналы">
    <Channels />
  </wired-tab>
  <wired-tab name="Two" label="Телепрограмма">
    <TvGuides />
  </wired-tab>
  <wired-tab name="Three" label="Загрузка">
    <TvGuidesUpload />
  </wired-tab>
</wired-tabs>
