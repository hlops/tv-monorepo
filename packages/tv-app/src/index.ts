import fastify from "fastify";
import fastifyCors from "fastify-cors";
import fastifyMultipart from "fastify-multipart";
import * as fs from "fs";
import * as http from "http";
import { Readable } from "stream";
import playlistRoute from "./routes/playlist";
import tvGuideRoute from "./routes/tvGuide";

const server = fastify<http.Server, http.IncomingMessage, http.ServerResponse>({});

server.register(fastifyCors, { methods: ["GET", "PUT", "POST", "FETCH"] });

interface BodyEntry {
  data: Buffer;
  filename: string;
  encoding: string;
  mimetype: string;
  limit: false;
  tempFileName?: string;
}

server.register(fastifyMultipart, {
  addToBody: true,
  sharedSchemaId: "MultipartFileType",
  onFile: (
    fieldName: string,
    stream: Readable,
    filename: string,
    encoding: string,
    mimetype: string,
    body: Record<string, BodyEntry>
  ) => {
    console.log(fieldName, filename, encoding, mimetype);
    body.file.tempFileName = "./node_data/uploadFile";
    const ws = fs.createWriteStream(body.file.tempFileName);
    //    stream.resume();
    stream.pipe(ws);
  }
  /*
    limits: {
      fieldNameSize: 100, // Max field name size in bytes
      fieldSize: 1000000, // Max field value size in bytes
      fields: 10, // Max number of non-file fields
      fileSize: 100, // For multipart forms, the max file size
      files: 1, // Max number of file fields
      headerPairs: 2000 // Max number of header key=>value pairs
    }
  */
});
server.register(playlistRoute);
server.register(tvGuideRoute);
server.register(tvGuideRoute);

const start = async () => {
  try {
    await server.listen(3000);
  } catch (err) {
    console.log(err);
    server.log.error(err);
    process.exit(1);
  }
};

/*
process.on("uncaughtException", error => {
  console.error(error);
});
process.on("unhandledRejection", error => {
  console.error(error);
});

*/
start();
