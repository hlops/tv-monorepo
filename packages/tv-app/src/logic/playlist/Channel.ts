import crypto from 'crypto';

export enum ChannelOptions {
	group = 'group-title'
}

export interface Channel {
	title: string;
	url: string;
	options: _.Dictionary<string>;
}

export class ChannelHelper {
	public static getHash(url: string): string {
		const md5sum = crypto.createHash('md5');
		return md5sum.update(url).digest('hex');
	}
}
