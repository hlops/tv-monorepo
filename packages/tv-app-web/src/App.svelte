<script>
  import MainTabs from "./MainTabs.svelte";
</script>

<main>
  <MainTabs />
</main>
