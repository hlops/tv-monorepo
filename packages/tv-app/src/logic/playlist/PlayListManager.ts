import keyFileStorage from 'key-file-storage';
import { KeyFileStorage } from 'key-file-storage/dist/src/key-file-storage';
import _ from 'lodash';
import { ChannelHelper } from './Channel';
import { M3uParser } from './m3u/M3uParser';

export class PlayListManager {
	constructor(private readonly storage: KeyFileStorage = keyFileStorage('node_data', true)) {
		this.storage = storage;
	}

	public importPlayList(stream: NodeJS.ReadableStream): Promise<void> {
		return new M3uParser().parse(stream).then(channels => {
			return Promise.all(
				_.map(
					channels,
					channel =>
						this.storage(`channels/${ChannelHelper.getHash(channel.url)}`, channel)
				)
			).then();
		});
	}
}
