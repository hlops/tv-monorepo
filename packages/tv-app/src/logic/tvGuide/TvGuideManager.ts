import keyFileStorage from 'key-file-storage';
import { KeyFileStorage } from 'key-file-storage/dist/src/key-file-storage';
import _ from 'lodash';
import { XmltvParser } from './xmltv/XmltvParser';

export class TvGuideManager {
	constructor(private readonly storage: KeyFileStorage = keyFileStorage('node_data', true)) {
		this.storage = storage;
	}

	public importXmltv(stream: NodeJS.ReadableStream): Promise<void> {
		return new XmltvParser().parse(stream).then(channels => {
			return Promise.all(
				_.map(channels, (channel, name) => this.storage(`guide/${name}`, channel))
			).then();
		});
	}
}
