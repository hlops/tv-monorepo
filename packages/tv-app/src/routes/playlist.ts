import { FastifyInstance } from "fastify";
import * as fs from "fs";
import * as _ from "lodash";
import { Channel, ChannelOptions } from "tv-common";
import { QueryParams } from "../logic/common/QueryParams";
import { ChannelHelper } from "../logic/playlist/Channel";
import { PlayListManager } from "../logic/playlist/PlayListManager";

function createChannelDto(channel: Channel) {
  return {
    id: ChannelHelper.getHash(channel.url),
    title: channel.title,
    url: channel.url,
    group: channel.options[ChannelOptions.group]
  };
}

export default (server: FastifyInstance, opts: { prefix: string }, next: (err?: Error) => void) => {
  server.route({
    url: "/playlist",
    method: ["GET"],
    handler: async (request, reply) => {
      console.log(opts);
      return new PlayListManager().find(new QueryParams(request.query)).then(channels =>
        reply.send({
          channels: _.map(channels, channel => createChannelDto(channel))
        })
      );
    }
  });

  server.post("/playlist", {
    handler: (request, reply) => {
      console.log("file:", request.body.file);
      new PlayListManager().import(fs.createReadStream(request.body.file.tempFileName)).then(() => {
        reply.send();
      });
    }
  });

  next();
};
