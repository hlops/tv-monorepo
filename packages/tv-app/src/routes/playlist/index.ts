import { FastifyInstance } from "fastify";
import * as fs from "fs";
import { IncomingMessage, Server, ServerResponse } from "http";
import { Readable } from "stream";
import {QueryParams} from "../../logic/common/QueryParams";
import { PlayListManager } from "../../logic/playlist/PlayListManager";

export default (
  server: FastifyInstance<Server, IncomingMessage, ServerResponse>,
  opts: { prefix: string },
  next: (err?: Error) => void
) => {
  server.route({
    url: "/playlist",
    method: ["GET", "HEAD"],
    handler: async (request, reply) => {
      return new PlayListManager()
        .findPlayLists(new QueryParams(request.query))
        .then(channels => reply.send({ channels }));
    }
  });

  server.post("/playlist", {
    handler: (request, reply) => {
      new PlayListManager()
        .importPlayList(fs.createReadStream(request.body.file.tempFileName))
        .then(() => {
          reply.send();
        });
    }
  });

  next();
};
