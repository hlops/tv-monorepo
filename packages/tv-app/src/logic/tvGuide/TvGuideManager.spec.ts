import fs from 'fs';
import keyFileStorage from 'key-file-storage';
import { KeyFileStorage } from 'key-file-storage/dist/src/key-file-storage';
import { PlayListManager } from '../playlist/PlayListManager';
import { TvGuideManager } from './TvGuideManager';

let storage: KeyFileStorage;

describe('TvGuideManager', () => {
	beforeAll(() => {
		storage = keyFileStorage('node_data/test', true);
	});

	afterAll(() => {
		delete storage['*'];
	});

	beforeEach(() => {
		delete storage['*'];
	});

	it('importPlayList() can successfully import channels', done => {
		new TvGuideManager(storage)
			.importXmltv(fs.createReadStream('./samples/xmltv-small.xml', 'utf8'))
			.then(() => {
				expect(storage['guide/']).toMatchSnapshot();
                expect(storage['guide/CBS Affiliate']).toMatchSnapshot();
				done();
			});
	});
});
