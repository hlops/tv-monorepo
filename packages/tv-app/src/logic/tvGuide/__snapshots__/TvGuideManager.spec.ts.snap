// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TvGuideManager importPlayList() can successfully import channels 1`] = `
Array [
  "guide/CBS Affiliate",
  "guide/PBS Affiliate",
]
`;

exports[`TvGuideManager importPlayList() can successfully import channels 2`] = `
Array [
  Object {
    "categories": Array [
      "Talk",
      "News",
      "Series",
    ],
    "desc": "Republican candidate John McCain; premiere of the film \\"The Dark Knight.\\"",
    "start": 2008-07-15T12:00:00.000Z,
    "stop": 2008-07-15T14:00:00.000Z,
    "title": "The Early Show",
  },
  Object {
    "categories": Array [
      "Talk",
      "Series",
    ],
    "desc": "Actresses Kim Raver, Brooke Shields and Lindsay Price (\\"Lipstick Jungle\\"); women in their 40s tell why they got breast implants; a 30-minute meal.",
    "start": 2008-07-15T14:00:00.000Z,
    "stop": 2008-07-15T15:00:00.000Z,
    "title": "Rachael Ray",
  },
  Object {
    "categories": Array [
      "Game show",
      "Series",
    ],
    "desc": "Contestants bid for prizes then compete for fabulous showcases.",
    "start": 2008-07-15T15:00:00.000Z,
    "stop": 2008-07-15T16:00:00.000Z,
    "title": "The Price Is Right",
  },
  Object {
    "categories": Array [
      "Game show",
      "Series",
    ],
    "desc": undefined,
    "start": 2008-07-15T16:00:00.000Z,
    "stop": 2008-07-15T16:30:00.000Z,
    "title": "Jeopardy!",
  },
  Object {
    "categories": Array [
      "Soap",
      "Series",
    ],
    "desc": "Jeff thinks Kyon stole the face cream; Nikki asks Nick to give David a chance; Amber begs Adrian to go to Australia.",
    "start": 2008-07-15T16:30:00.000Z,
    "stop": 2008-07-15T17:30:00.000Z,
    "title": "The Young and the Restless",
  },
]
`;
